### PR TITLE
Border Clipping

### PR DESCRIPTION
**Pull request by danwalmsley**
## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

File: src/Avalonia.Visuals/Rendering/SceneGraph/SceneBuilder.cs [hunk 1]
```diff
@@ -201,6 +205,7 @@ private static void Update(DrawingContext context, Scene scene, VisualNode node,
                     node.ClipBounds = clipBounds;
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Rendering/SceneGraph/SceneBuilder.cs:205]: @grokys there are some non-trivial calculations that are applied to `clipBounds` here, so I'm not sure how well my change would work with transforms and such.

**Comment by Gillibald** [comment]
Is a geometry clip really that slow? Is there some data that justifies all of this specialization? Also why don't we use a geometry implementation that wraps the rounded rect drawing call?

This could be entirely written in the platform code.

**Comment by kekekeks** [comment]
Using geometry clip would conflict with the existing `Clip` property, so we can't reuse it.

Using specialized SKRRect overload is faster than SKPath. That's exactly why Skia has specialized clipping for SRRect.

**Comment by Gillibald** [comment]
Can't we introduce a new property `public IGeometryImpl VisualClip { get; protected set; }`?

Clip has priority if it is defined.

By default, this returns just a lightweight `RectangleImpl` that uses optimized code path

**Comment by robloo** [comment]
This might be a good time to align to WPF instead of UWP. UWP unnecessarily sacrificed some power here by not allowing rectangular clipping.

**UIElement.Clip**

Both WPF and UWP have UIElement.Clip properties. However, WPF can take any Geometry allowing for non-rectangular clipping. UWP can only use a RectangleGeometry for clipping.

 * WPF: public Geometry UIElement.Clip
 * UWP: public RectangleGeometry UIElement.Clip

**UIElement.ClipToBounds**

In WPF it's possible to clip child contents to the parents bounds by setting ClipToBounds to True. UWP doesn't have this property at all. The work-around is to use UIElement.Clip which can only do rectangular clipping.

**Comment by jmacato** [comment]
@robloo we already have the clip property for all elements i think. idk about the cliptobounds, iirc we have that property but it doesnt work...

**Comment by robloo** [comment]
@jmacato Ok, apologies for misunderstanding. I see that you are already following WPF which is great! I also see this is related to rendering specifically so I clearly need to read things through more before commenting.

http://reference.avaloniaui.net/api/Avalonia.Controls/Control/#Properties

**Comment by Gillibald** [comment]
The difference is that we have the Clip property on Visual rather than on StyledElement(UIElement)

File: src/Avalonia.Visuals/Rendering/SceneGraph/SceneBuilder.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Visuals/Rendering/SceneGraph/SceneBuilder.cs:205]: Looks good to me and just tested it in the `RenderDemo` with a rotation transform animation: all looks to me good.

**Comment by danwalmsley** [comment]
image

**Pull request merged**